kratos/containers/variables_list.h

```diff
@@ -69,532 +35,433 @@ SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+		{
+			mDataSize = rOther.mDataSize;
+			mHashFunctionIndex = rOther.mHashFunctionIndex;
+			mKeys = rOther.mKeys;
+			mPositions = rOther.mPositions;
+			mVariables = rOther.mVariables;
+
+			return *this;
+		}
+
+		IndexType operator()(IndexType VariableKey) const
+		{
+			return GetPosition(VariableKey);
+		}
+
+		template<class TDataType>
+		IndexType operator()(Variable<TDataType> const& ThisVariable) const
+		{
+			return GetPosition(ThisVariable.Key());
+		}
+
+		const VariableData* operator[](IndexType Index) const
+		{
+			return mVariables[Index];
+		}
+
+
+		bool operator==(const VariablesList& r) const // nothrow
+		{
+			if (size() != r.size())
+				return false;
+			else
+				return std::equal(mPositions.begin(), mPositions.end(), r.mPositions.begin()) &&
+				std::equal(mVariables.begin(), mVariables.end(), r.mVariables.begin());
+		}
+
+
+		///@}
+		///@name Operations
+		///@{
+
+		const_iterator             begin() const
+		{
+			return const_iterator(mVariables.begin());
+		}
+		const_iterator             end() const
+		{
+			return const_iterator(mVariables.end());
+		}
+		const_reverse_iterator     rbegin() const
+		{
+			return const_reverse_iterator(mVariables.rbegin());
+		}
+		const_reverse_iterator     rend() const
+		{
+			return const_reverse_iterator(mVariables.rend());
+		}
+		ptr_const_iterator         ptr_begin() const
+		{
+			return mVariables.begin();
+		}
+		ptr_const_iterator         ptr_end() const
+		{
+			return mVariables.end();
+		}
+		ptr_const_reverse_iterator ptr_rbegin() const
+		{
+			return mVariables.rbegin();
+		}
+		ptr_const_reverse_iterator ptr_rend() const
+		{
+			return mVariables.rend();
+		}
+
+		const_reference  front() const /* nothrow */
+		{
+			assert(!IsEmpty());
+			return *mVariables.front();
+		}
+		const_reference  back() const  /* nothrow */
+		{
+			assert(!IsEmpty());
+			return *mVariables.back();
+		}
+
+		size_type size() const
+		{
+			return mVariables.size();
+		}
+
+		size_type max_size() const
+		{
+			return mVariables.max_size();
+		}
+
+		void swap(VariablesList& rOther)
+		{
+			SizeType temp = mDataSize;
+			mDataSize = rOther.mDataSize;
+			rOther.mDataSize = temp;
+
+			temp = mHashFunctionIndex;
+			mHashFunctionIndex = rOther.mHashFunctionIndex;
+			rOther.mHashFunctionIndex = temp;
+
+			mVariables.swap(rOther.mVariables);
+			mKeys.swap(rOther.mKeys);
+			mPositions.swap(rOther.mPositions);
+		}
+
+		template<class TOtherDataType>
+		void push_back(TOtherDataType const& x)
+		{
+			Add(x);
+		}
+
+		void clear()
+		{
+			mDataSize = 0;
+			mHashFunctionIndex = 0;
+			mVariables.clear();
+			mKeys.clear();
+			mPositions.clear();
+		}
+
+
+		void Add(VariableData const& ThisVariable)
+		{
+			if (ThisVariable.Key() == 0)
+				KRATOS_THROW_ERROR(std::logic_error,
+					"Adding uninitialize variable to this variable list. Check if all variables are registered before kernel initialization", "");
+
+			if (Has(ThisVariable))
```

Comment:
what happens if the hash happens to be the same? i understand it is rare, but it might happen, so i guess this would be the place to check that the Variable is effectively the same and otherwise throw an error